### ISSUE TITLE
receive_message_chain ... exactly() || twice() || <similar>() produces undefined method

### ISSUE THREAD

**Issue by brycemcd** [rspec/rspec-mocks#921]
I'm testing a piece of code in a rails controller like:

```ruby
things.each do |thing|
  current_user.all_the_things.create(thing)
end

render json: {success: true}
```

I have a mock like:

```ruby
allow_any_instance_of(User).to receive_message_chain('all_the_things.create').
  exactly(2).times.and_return(true)
# ... more setup
expect(response).to be_ok
```

When the spec is run, an exception is thrown:
`undefined method 'exactly' for for #RSpec::Mocks::Matchers::ReceiveMessageChain:0x0000000afb3d08>`

The same happens if `exactly` is replaced with `twice` or [anything else defined in the documentation](https://relishapp.com/rspec/rspec-mocks/docs/setting-constraints/receive-counts)

I'm happy to provide a more concrete test case (and pull request), but before I get too deep into it I want to confirm that the behavior I'm expecting is rational.

According to [the message chain docs](https://relishapp.com/rspec/rspec-mocks/v/3-2/docs/working-with-legacy-code/message-chains), the `receive_message_chain` method should be a replacement for the `receive` message. My expectation is that the `receive_message_chain` message should be allowed to have the 'receive count' messages chained to them in the same way that `receive` does.

All that said, this whole approach feels a bit like a code smell, but I'm not confident of another approach. Law of Demeter would suggest not chaining `current_user.all_the_things` to `create`. However, it's good practice in rails to create relationships belonging to the user by sending messages to related objects through the original object (`current_user` in this case).

TL;DR: is the receive_message_chain behavior described above a bug or a feature?

**Comment by myronmarston** [comment]
It's by design. `receive_message_chain` isn't intended to support the full level of customization that you get with `receive`.

The docs could perhaps be improved to make that more clear -- want to open a PR to improve that?

**Comment by brycemcd** [comment]
yeah, happy to open a PR.

What is the intended support of `receive_message_chain` vs. `receive` ?

The more I noodle on the use case above, the more I think the 'receive count' messages make sense.

**Comment by myronmarston** [comment]
I was going to say that `receive_message_chain` is only intended to support configuration the response (via `and_return`, `and_yield`, etc) but it looks like we added support for `with` to #697.  We did that for `stub_chain` parity, but it's actually a bit confusing -- `with` specifies the expected arguments for _one_ method call but `receive_message_chain` is about a _chain_ of method calls so there's a bit of a semantic mis-match.  Looks like `with` applies only to the last message in the chain, which is kinda weird.  My guess is that `stub_chain` simply returned the last stub in the chain, allowing further customizations to be set on the last one as opposed to that being intentionally planned out originally.

Given the simplicity of the implementation of #697 it could make sense to add support for the constraints to `receive_message_chain`, although it's still kinda semantically weird in terms of one method call vs a chain of them.

@JonRowe, what do you think?

**Comment by JonRowe** [comment]
I think if `stub_chain` didn't support it then we don't add it for `receive_messages`, I only added `with` for parity's sake because it was easy (I'd have preferred not supporting it really)...

**Comment by brycemcd** [comment]
Is it fair to say the biggest issue is supporting `with` when chaining `stub_chain` or `receive_message_chain`? I definitely see that it creates ambiguity in which method the arguments are passed to (although there _is_ a convention for it). It probably makes writing a code smell like that too easy.

On the other hand, it is a good practice to create objects owned by another object (`current_user.bank_accounts` for ex) using the method chain. Considering a world where the `stub_chain` or `receive_message_chain` does not support being able to chain other messages like `with` or `twice`, I think the alternative to the example code above would be:

```ruby
things.each do |thing|
  att = AllTheThing.new(thing)
  att.user = current_user
  att.save
end
```

in order to create mocks that accept more than one expectation. There's nothing wrong with that code per se, but it does require changing the implementation code in order to satisfy the needs of the testing harness. That doesn't seem quite right.

If `with` is the message that is the most problematic then it probably makes sense to deprecate that specific method in a future version.

At any rate, I'm happy to submit a PR for a documentation update but it would be helpful to have more insight into the design decision.

**Comment by JonRowe** [comment]
The usage of `stub_chain` and `receive_message_chain` is considered a test smell, thus we try not to encourage their adoption by extending their feature set.

**Comment by brycemcd** [comment]
Agreed, but in the case of the `current_user.bank_accounts` the smell is just a smell and not a bad practice. Shouldn't the framework support the accepted practice? Otherwise, I totally agree that you wouldn't want to provide something that's too easy to create a bad practice

**Comment by JonRowe** [comment]
Well no, because we have a better solution (even solely from a test perspective) instead of using `receive_message_chain` to provide the end solution use it to provide the last but one and provide a normal `instance_double` for that last value.

In the case of a simple chain like `current_user.bank_accounts` you don't even need the chain at all, it would be better to stub `current_user` as an `instance_double` (of assumedly `User`) and then place expectations on that double directly, gaining not only the full expectation feature set, but also the benefits of verifying doubles.

**Comment by myronmarston** [comment]
``` ruby
things.each do |thing|
  att = AllTheThing.new(thing)
  att.user = current_user
  att.save
end
```

On its own, this code is fine.  However, once you're at a point where you have a need to substitute a test double for it, you run into the complexity of the interaction here.  This bit of code has to use 3 method calls (`new`, `user=` and `save`) when it could be one, simpler interface:

``` ruby
class AllTheThing
  def self.update_user(thing, user)
    att = AllTheThing.new(thing)
    att.user = user
    att.save
  end
end

things.each do |thing|
  AllTheThing.update_user(thing, current_user)
end
```

Now you've created a very simple seam (`AllTheThing.update_user`) which can be very simply mocked w/o the need to resort to `receive_message_chain`.  IMO, this makes the code better: you've reified a common use case into a first-class method that directly supports that use case with a simpler interface.

We're hesitant to extend the feature set of `receive_message_chain` because there's good design pressure that comes from a complex interaction in your implementation code forcing a complex use of your mocking tool.  You can, of course, mock this just fine w/o `receive_message_chain`:

``` ruby
alt = instance_double(AllTheThing)
expect(alt).to receive(:user=).with(current_user)
expect(alt).to receive(:save)
allow(AllTheThing).to receive(:new).with(thing).and_return(alt)
```

...but that rightfully makes you uncomfortable.  It's a complex interaction that you are encoding directly into your tests.  Mocking tools should exert design pressure towards _simpler interfaces_ rather than simply encoding complex interactions in your tests.  Using `receive_message_chain` hides the complexity of the interaction (which can be OK in some cases, particularly when getting existing legacy code under test), but isn't generally a good idea.  Instead of looking to solve the complexity of mocking this interaction with an increased feature set from rspec-mocks, consider if there's a way you can improve the implementation by simplifying the interaction.  Usually there is.

Anyhow, I'm personally not opposed to extending the feature set of `receive_message_chain`, provided that:

* The implementation is very simple -- none of the core team members actively use the feature and we don't want to be stuck supporting something complex we don't use.
* There are no semantic ambiguities in the API
* Someone else does the work and submits the PR (we have other things we care more about that we are working).

Does that help explain our position (or at least my position -- @JonRowe may feel different)?

**Comment by JonRowe** [comment]
Personally I'm still opposed to extending the feature set of this, I'd like us to one day deprecate it entirely.

**Comment by brycemcd** [comment]
I got caught up in other things this weekend, but I should be able to submit a doc update later this week. I think I've got my head around the design decision enough to do the PR. It likely won't be a profound PR, but the information in this thread is really helpful to understanding how to use `receive_message_chain` vs. `double`

Much appreciated and I hope it helps others

**Comment by samphippen** [comment]
@brycemcd are you still interested in doing this? We love it when people help us with our documentation ^_^

**Comment by brycemcd** [comment]
@samphippen - it totally slipped my mind. I'm interested and I've blocked out some time to get it done by the end of the week.

**Comment by JonRowe** [comment]
Fixed by #984

**Comment by Bartuz** [comment]
@myronmarston do you consider using `receive_message_chain` to stub chained scopes on model as code-smell?

**Comment by myronmarston** [comment]
Yes.

**Comment by Bartuz** [comment]
So how should I write expectation for scopes? Does the example below make any sense? I have the idea to add new spy for each chained scope and than expectation on it

```
let(:relation) { spy 'relation' }

before do 
    expect(Model).to receive(:with_color).with('red') { relation }
   expect(relation).to receive(:with_shape).with(:triangle) { [instance_double('model')] }
end

it 'does cool things' do
   expect(subject).to eq ...
end
```

I've tried to connect scopes together (like `scope :with_color_and_shape, ->(color,share) {  ... }`) so I could just do `expect(Model).to receive(:with_color_and_shape).with(color, shape) {}` but I ended up with huge scope taking many arguments and calling bunch of smaller one which was hard to manage.

**Comment by myronmarston** [comment]
Remember that a "code smell" isn't a rule and for nearly every code smell, there's a situation for which the trade off to use the smelly code makes sense.  I don't have any good advice for rails because I'm not a rails user and haven't been since Rails 2.

**Comment by JonRowe** [comment]
@Bartuz the common rule of thumb would be to combine the scopes at the model level, but this isn't always going to to be pretty either as you've found out. The only real way to resolve this "smell" is to refactor your design to not need to reach in across so many scopes, but I appreciate that's sometimes not possible / desirable. As @myronmarston says there are sometimes exceptions to rules and as long as you can justify to your team you can decide this is such an exception :)
